### DISCUSSION TITLE
Does the shell itself expand globs or is glob-expansion the job of applications?

### DISCUSSION THREAD

**Discussion by wmstack** [nushell/nushell#6066] [Q&A]
I noticed that `echo **` produces `**` instead of every possible subdirectory. Still I want to make sure.

**Comment by panicbit** [comment]
I'm not sure what the intended goal of nushell is regarding globs. Some commands like `ls` do glob expansion themself, but there's also a command (`glob`) to do it explicitly:

![grafik](https://user-images.githubusercontent.com/628445/184551350-fb921e0a-edf9-44a3-b24b-e2432fb53575.png)

**Reply by fdncred** [reply]
The glob command was meant to help with proper globbing since some of nu-glob was modified to make it more functional for ls.

**Comment by panicbit** [comment]
To be honest, I'd probably prefer it if `ls` did not do globbing, especially if there is no way for opting out of it on a case-by-case basis.
Imagine passing paths from somewhere else into `ls` with some files actually containing `*` or `?`.
That would cause those files to not be found by `ls` because it would erroneously be trying to do globbing.

**Reply by fdncred** [reply]
i'd love to have the glob command and the ls command work better together like `glob blah | ls` or `ls (glob blah)`. however, i think removing globbing from ls in general would probably be a bad idea because everyone expects ls to do some type of globbing.

**Reply by Taywee** [reply]
I'm in support of one of the following:

* Glob patterns are evaluated directly by the shell.
  * Similar to the way it works in Bourne-derivative shells, with magic characters that evaluate to a glob (or, to avoid confusing interactions with `*` being multiply in some contexts and glob in others, maybe some glob-specific string prefix: `*'file-*-name'` or something to indicate that it's a glob)
  * This would require a glob being evaluated to a list or even a special "glob" type.
* Just use `glob` and have people adjust to the idea of not having automatic globbing.

I don't think command should be doing their own globbing.  It leads to surprising behavior like some commands unexpectedly not globbing, or for plugins, having subtly different globbing behavior for plugins than the shell itself.

Personally, I don't like my shell doing my globbing for me either.  I'd rather anything that operates on filenames (`cp`, `ls`, `touch`, `rm`, etc) be able to take in a list as the source/target argument.  It also opens up some interesting opportunities for nu-specific functionality that I've long wished other tools could easily do, like having `cp [alpha beta] [gamma delta]` copy alpha and beta into multiple destination directories.

**Reply by amtoine** [reply]
`glob blah | ls` and `ls (glob blah)` are quite interesting ideas, i wouldn't mind if `ls` was not doing globbing :yum:
